lib/kube/proxy/forwarder.go

```diff
@@ -652,16 +664,66 @@ func (f *Forwarder) setupForwardingHeaders(ctx *authContext, sess *clusterSessio
 	req.Header.Add("X-Forwarded-Host", req.Host)
 	req.Header.Add("X-Forwarded-Path", req.URL.Path)
 
+	return nil
+}
+
+// setupImpersonationHeaders sets up Impersonate-User and Impersonate-Group headers
+func setupImpersonationHeaders(log log.FieldLogger, ctx *authContext, headers http.Header, bearerToken string) error {
+	var impersonateUser string
+	var impersonateGroups []string
+	for header, values := range headers {
+		if !strings.HasPrefix(header, "Impersonate-") {
+			continue
+		}
+		switch header {
+		case ImpersonateUserHeader:
+			if impersonateUser != "" {
+				return trace.AccessDenied("%v, user already specified to %q", ImpersonationRequestDeniedMessage, impersonateUser)
+			}
+			if len(values) == 0 || len(values) > 1 {
+				return trace.AccessDenied("%v, invalid user header %q", ImpersonationRequestDeniedMessage, values)
+			}
+			impersonateUser = values[0]
+			if _, ok := ctx.kubeUsers[impersonateUser]; !ok {
+				return trace.AccessDenied("%v, user header %q is not allowed in roles", ImpersonationRequestDeniedMessage, impersonateUser)
+			}
```

Comment:
Will this reject a user attempting to impersonate their own name?  If so, this is potentially confusing since users are allowed to impersonate themselves as a default behavior.

Reply:
Their own name could mean something else in the context of the Kubernetes cluster, e.g. IAM role with some privileges. That's why impersonation is fully controlled by roles and there are no default allowed values.

Reply:
> That's why impersonation is fully controlled by roles and there are no default allowed values.

[This line](https://github.com/gravitational/teleport/blob/4ae31f7ca1a69cc218b9ba424d776afef1d9998a/lib/kube/proxy/forwarder.go#L711) looks like it causes users to impersonate their own usernames by default, regardless of what the roles say.

Reply:
Got it, thanks for pointing it out, that was a behavior before `kuberenetes_users`, so I tried to preserve the default behavior of the system, but now I see how it can be confusing. I wonder if we should always allow users to impersonate themselves then :thinking:

Reply:
I don't know the ramifications of kubernetes user impersonation well enough to comment on whether self-impersonation is a *desirable* default... I will say that I prefer it over breaking existing deployments (though I think it should either always or never be the default).

If we don't want to keep self-impersonation as the default, perhaps we could provide a migration which allowed existing deployments to keep their behavior?  E.G.:

```yaml
options:
  k8s_user_passthrough: yes
```

Reply:
@fspmarshall I think this will be reasonable but a bit confusing behavior to explain to users and admins:

* By default, if no `kubernetes_users` is set (which will be a majority), user will impersonate themselves
* As long as at least one `kubernetes_users` is set, it will start controlling the overall list of users allowed by the client to impersonate. If it does not include actual user name, well, will be rejected (otherwise there will be no way to exclude the user from the list)
* If the `kuberentes_users` set includes only one user (many times that's the real intent), teleport will default to it, otherwise it will refuse to select. This will enable `IAM#{{external.email}}` use case that is the driver for this feature.

Reply:
@klizhentas sounds good, that does feel easier to explain.